### ISSUE TITLE
Cgroups are not restored after restoring container from checkpoint

### ISSUE THREAD

**Issue by tswift242** [moby/moby#32442]
**Description**

See https://github.com/xemul/criu/issues/305

According to @avagin, this seems to be a regression in runc. I'm opening this issue to track getting the runc fix into docker.

**Output of `docker version`:**

```
Client:
 Version:      17.03.0-ce
 API version:  1.26
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:57:58 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.0-ce
 API version:  1.26 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   3a232c8
 Built:        Tue Feb 28 07:57:58 2017
 OS/Arch:      linux/amd64
 Experimental: true
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 4
Server Version: 17.03.0-ce
Storage Driver: overlay
 Backing Filesystem: extfs
 Supports d_type: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host ipvlan macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 977c511eda0925a723debdc94d09459af49d082a
runc version: a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70
init version: 949e6fa
Security Options:
 apparmor
Kernel Version: 4.10.5-041005-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 14.94 GiB
Name: ip-10-97-3-17
ID: G3FR:5WAN:LTDA:O5EK:HAOA:3RJG:JKU5:HBDD:TTG5:GASD:PLMB:CFIK
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: true
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
Following that discussion; this was the commit that caused the change in behaviour; https://github.com/opencontainers/runc/commit/8a740d5391a7523a26ddb886a823cadcbf70ac6d

/cc @mlaventure

**Comment by thaJeztah** [comment]
Hm, but looks like docker 17.03.0-ce does not have that commit . RunC version for 17.03.0 is 

https://github.com/docker/runc/commit/a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70, and (for example) [libcontainer/cgroups/fs/pids.go#L21-L270(https://github.com/docker/runc/blob/a01dafd48bc1c7cc12bdb01206f9fea7dd6feb70/libcontainer/cgroups/fs/pids.go#L21-L27) does not contain the changes from that commit

**Comment by tswift242** [comment]
@avagin @mlaventure Any thoughts on @thaJeztah's comment above?

**Comment by avagin** [comment]
@thaJeztah I don't understand what you mean.

> Hm, but looks like docker 17.03.0-ce does not have that commit . RunC version for 17.03.0 is docker/runc@a01dafd, and (for example) libcontainer/cgroups/fs/pids.go#L21-L27 does not contain the changes from that commit

But if we look at the code and the patch, we can see that Set isn't called form Apply, so the code contains changes from the patch.

```
func (s *PidsGroup) Apply(d *cgroupData) error {
	_, err := d.join("pids")
	if err != nil && !cgroups.IsNotFound(err) {
		return err
	}
	return nil
}
```

```
@@ -18,15 +18,10 @@ func (s *PidsGroup) Name() string {
 }
 
 func (s *PidsGroup) Apply(d *cgroupData) error {
-       dir, err := d.join("pids")
+       _, err := d.join("pids")
        if err != nil && !cgroups.IsNotFound(err) {
                return err
        }
-
-       if err := s.Set(dir, d.config); err != nil {
-               return err
-       }
-
        return nil
 }
```

**Comment by thaJeztah** [comment]
@mlaventure ^^

**Comment by avagin** [comment]
This issue was fixed in runc: https://github.com/opencontainers/runc/commit/57ef30a2ae432115f04678bc905b59451bddb5b8

**Comment by avagin** [comment]
How often is runc updated in Docker?

**Comment by cpuguy83** [comment]
@avagin As need for each release.

**Comment by mlaventure** [comment]
Michael is working on updating `containerd` to use `runc` master, this will then include this fix

**Comment by mlaventure** [comment]
runc was updated, this should be fixed now thanks to #33007
